File: .claude/hooks/moai/steering_guard.py
#!/usr/bin/env python3
"""UserPromptSubmit guard enforcing steering safety rules with session notifications."""

from __future__ import annotations

import json
import re
import sys
import os
from pathlib import Path
from typing import Dict

BANNED_PATTERNS = (
    (re.compile(r'(?i)ignore (the )?(claude|constitution|steering|instructions)'), '헌법/지침 무시는 허용되지 않습니다.'),
    (re.compile(r'(?i)disable (all )?(hooks?|guards?|polic(y|ies))'), 'Hook/Guard 해제 요청은 차단되었습니다.'),
    (re.compile(r'(?i)rm -rf'), '위험한 셸 명령을 프롬프트로 제출할 수 없습니다.'),
    (re.compile(r'(?i)drop (all )?safeguards'), '안전장치 제거 요청은 거부됩니다.'),
    (re.compile(r'(?i)clear (all )?(memory|steering)'), 'Steering 메모리를 강제 삭제하는 요청은 지원하지 않습니다.'),
)
# 세션별 알림 상태 추적을 위한 임시 파일
SESSION_NOTIFIED_FILE = "/tmp/moai_session_notified"

def _check_moai_project() -> bool:
    """MoAI 프로젝트 여부 확인"""
    current_dir = Path.cwd()
    return (current_dir / ".moai").exists() and (current_dir / "CLAUDE.md").exists()

def _show_session_notice() -> None:
    """세션 시작 알림 표시 (최초 1회)"""
    if os.path.exists(SESSION_NOTIFIED_FILE):
        return  # 이미 알림을 표시했음

    if not _check_moai_project():
        return  # MoAI 프로젝트가 아님

    # 알림 표시
    print("🚀 MoAI-ADK 프로젝트가 감지되었습니다!", file=sys.stderr)
    print("📖 개발 가이드: CLAUDE.md | TRUST 원칙: .moai/memory/development-guide.md", file=sys.stderr)
    print("⚡ 워크플로우: /moai:1-spec → /moai:2-build → /moai:3-sync", file=sys.stderr)
    print("🔧 디버깅: /moai:4-debug | 설정 관리: @agent-cc-manager", file=sys.stderr)
    print("", file=sys.stderr)

    # 알림 완료 표시
    try:
        with open(SESSION_NOTIFIED_FILE, "w") as f:
            f.write("notified")
    except:
        pass  # 임시 파일 생성 실패는 무시

def _load_input() -> Dict[str, object]:
    try:
        return json.load(sys.stdin)
    except json.JSONDecodeError as exc:
        print(f"ERROR steering_guard: 잘못된 JSON 입력 ({exc})", file=sys.stderr)
        sys.exit(1)


def main() -> None:
    # 세션 시작 알림 확인 (첫 번째 프롬프트에서만)
    _show_session_notice()

    data = _load_input()
    prompt = data.get('prompt')
    if not isinstance(prompt, str):
        sys.exit(0)

    for pattern, message in BANNED_PATTERNS:
        if pattern.search(prompt):
            print(f'BLOCKED: {message}', file=sys.stderr)
            print('HINT: CLAUDE.md와 @.moai/project/* 문서를 기반으로 목표/제약을 명시해 주세요.', file=sys.stderr)
            sys.exit(2)

    # Provide lightweight steering context back to Claude.
    print('Steering Guard: 개발 가이드과 TAG 규칙을 준수하며 작업을 진행합니다.', flush=True)
    sys.exit(0)


if __name__ == '__main__':
    main()

File: moai-adk-ts/src/claude/hooks/security/steering-guard.ts
/**
 * @file steering-guard.ts
 * @description Steering guard hook for enforcing safety rules with session notifications
 * @version 1.0.0
 * @tag @SEC:STEERING-GUARD-013
 */

import type { HookInput, HookResult, MoAIHook, SecurityPattern } from '../types';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

/**
 * Banned patterns that should be blocked
 */
const BANNED_PATTERNS: SecurityPattern[] = [
  {
    pattern: /ignore (the )?(claude|constitution|steering|instructions)/i,
    message: '헌법/지침 무시는 허용되지 않습니다.',
    severity: 'critical',
  },
  {
    pattern: /disable (all )?(hooks?|guards?|polic(y|ies))/i,
    message: 'Hook/Guard 해제 요청은 차단되었습니다.',
    severity: 'critical',
  },
  {
    pattern: /rm -rf/i,
    message: '위험한 셸 명령을 프롬프트로 제출할 수 없습니다.',
    severity: 'high',
  },
  {
    pattern: /drop (all )?safeguards/i,
    message: '안전장치 제거 요청은 거부됩니다.',
    severity: 'critical',
  },
  {
    pattern: /clear (all )?(memory|steering)/i,
    message: 'Steering 메모리를 강제 삭제하는 요청은 지원하지 않습니다.',
    severity: 'high',
  },
];

/**
 * Session notification file path
 */
const SESSION_NOTIFIED_FILE = path.join(os.tmpdir(), 'moai_session_notified');

/**
 * Hybrid system status
 */
interface HybridStatus {
  status: 'full_hybrid' | 'python_only' | 'typescript_only' | 'legacy';
  description: string;
}

/**
 * Steering Guard Hook - TypeScript port of steering_guard.py
 */
export class SteeringGuard implements MoAIHook {
  name = 'steering-guard';

  async execute(input: HookInput): Promise<HookResult> {
    // Show session notice on first execution
    this.showSessionNotice();

    const prompt = input.prompt;
    if (!prompt || typeof prompt !== 'string') {
      return { success: true };
    }

    // Check against banned patterns
    for (const pattern of BANNED_PATTERNS) {
      if (pattern.pattern.test(prompt)) {
        return {
          success: false,
          blocked: true,
          message: pattern.message,
          exitCode: 2,
        };
      }
    }

    // Provide lightweight steering context
    return {
      success: true,
      message:
        'Steering Guard: 개발 가이드과 TAG 규칙을 준수하며 작업을 진행합니다.',
    };
  }

  /**
   * Check if this is a MoAI project
   */
  private checkMoAIProject(): boolean {
    const currentDir = process.cwd();
    const moaiPath = path.join(currentDir, '.moai');
    const claudePath = path.join(currentDir, 'CLAUDE.md');

    return fs.existsSync(moaiPath) && fs.existsSync(claudePath);
  }

  /**
   * Check hybrid system status
   */
  private checkHybridSystemStatus(): HybridStatus {
    const currentDir = process.cwd();

    // Check TypeScript project
    const tsProject = path.join(currentDir, 'moai-adk-ts');
    const hasTypeScript =
      fs.existsSync(tsProject) &&
      fs.existsSync(path.join(tsProject, 'package.json'));

    // Check Python bridge
    const pythonBridge = path.join(
      currentDir,
      'src',
      'moai_adk',
      'core',
      'bridge'
    );
    const hasPythonBridge =
      fs.existsSync(pythonBridge) &&
      fs.existsSync(path.join(pythonBridge, 'typescript_bridge.py'));

    if (hasTypeScript && hasPythonBridge) {
      return {
        status: 'full_hybrid',
        description: 'Python + TypeScript 완전 통합 🔗',
      };
    } else if (hasPythonBridge) {
      return {
        status: 'python_only',
        description: 'Python 브릿지 (TypeScript 없음) 🐍',
      };
    } else if (hasTypeScript) {
      return {
        status: 'typescript_only',
        description: 'TypeScript (브릿지 없음) ⚠️',
      };
    } else {
      return {
        status: 'legacy',
        description: '기존 Python 시스템 📦',
      };
    }
  }

  /**
   * Show session notice (first time only)
   */
  private showSessionNotice(): void {
    if (fs.existsSync(SESSION_NOTIFIED_FILE)) {
      return; // Already notified
    }

    if (!this.checkMoAIProject()) {
      return; // Not a MoAI project
    }

    // Check hybrid system status
    const hybridStatus = this.checkHybridSystemStatus();

    // Show notification
    console.error('🚀 MoAI-ADK 하이브리드 프로젝트가 감지되었습니다!');
    console.error(
      '📖 개발 가이드: CLAUDE.md | TRUST 원칙: .moai/memory/development-guide.md'
    );
    console.error(
      '⚡ 하이브리드 워크플로우: /moai:1-spec → /moai:2-build → /moai:3-sync'
    );
    console.error(`🔗 시스템 상태: ${hybridStatus.description}`);
    console.error('🔧 디버깅: /moai:4-debug | 설정 관리: @agent-cc-manager');
    console.error('');

    // Mark as notified
    try {
      fs.writeFileSync(SESSION_NOTIFIED_FILE, 'notified');
    } catch {
      // Ignore write errors
    }
  }
}

/**
 * CLI entry point for Claude Code compatibility
 */
export async function main(): Promise<void> {
  try {
    const { parseClaudeInput, outputResult } = await import('../index');
    const input = await parseClaudeInput();
    const guard = new SteeringGuard();
    const result = await guard.execute(input);
    outputResult(result);
  } catch (error) {
    console.error(
      `ERROR steering_guard: ${error instanceof Error ? error.message : 'Unknown error'}`
    );
    process.exit(1);
  }
}

// Execute if run directly
if (require.main === module) {
  main().catch(error => {
    console.error(
      `ERROR steering_guard: ${error instanceof Error ? error.message : 'Unknown error'}`
    );
    process.exit(1);
  });
}
